src/parquet/column/reader.h

```diff
@@ -155,50 +155,59 @@ inline size_t TypedColumnReader<TYPE>::ReadValues(size_t batch_size, T* out) {
-    return 0;
-  }
-
-  // TODO(wesm): keep reading data pages until batch_size is reached, or the
-  // row group is finished
-  batch_size = std::min(batch_size, num_buffered_values_);
-
-  size_t num_def_levels = 0;
-  size_t num_rep_levels = 0;
-
-  size_t values_to_read = 0;
-
-  // If the field is required and non-repeated, there are no definition levels
-  if (descr_->max_definition_level() > 0) {
-    num_def_levels = ReadDefinitionLevels(batch_size, def_levels);
-    // TODO(wesm): this tallying of values-to-decode can be performed with better
-    // cache-efficiency if fused with the level decoding.
-    for (size_t i = 0; i < num_def_levels; ++i) {
-      if (def_levels[i] == descr_->max_definition_level()) {
-        ++values_to_read;
+  *values_read = 0;
+  size_t total_values = 0;
+  int32_t level_offset = 0;
+  int32_t value_offset = 0;
+  size_t current_values_read = 0;
+  while (true) {
+    // HasNext invokes ReadNewPage
+    // Continue reading until there are no new pages
+    // or the batch_size is exhausted
+    if (!HasNext() || batch_size == 0) {
+      break;
+    }
+    int current_batch_size = std::min(batch_size, num_buffered_values_);
+    size_t num_def_levels = 0;
+    size_t num_rep_levels = 0;
+    size_t values_to_read = 0;
+
+    // If the field is required and non-repeated, there are no definition levels
+    if (descr_->max_definition_level() > 0) {
+      num_def_levels = ReadDefinitionLevels(current_batch_size,
+          def_levels + level_offset);
+      // TODO(wesm): this tallying of values-to-decode can be performed with better
+      // cache-efficiency if fused with the level decoding.
+      for (size_t i = 0; i < num_def_levels; ++i) {
+        if (def_levels[i + level_offset] == descr_->max_definition_level()) {
+          ++values_to_read;
+        }
       }
+    } else {
+      // Required field, read all values
+      values_to_read = current_batch_size;
     }
-  } else {
-    // Required field, read all values
-    values_to_read = batch_size;
-  }
 
-  // Not present for non-repeated fields
-  if (descr_->max_repetition_level() > 0) {
-    num_rep_levels = ReadRepetitionLevels(batch_size, rep_levels);
-    if (num_def_levels != num_rep_levels) {
-      throw ParquetException("Number of decoded rep / def levels did not match");
+    // Not present for non-repeated fields
+    if (descr_->max_repetition_level() > 0) {
+      num_rep_levels = ReadRepetitionLevels(current_batch_size,
+          rep_levels + level_offset);
+      if (num_def_levels != num_rep_levels) {
+        throw ParquetException("Number of decoded rep / def levels did not match");
+      }
     }
-  }
-
-  *values_read = ReadValues(values_to_read, values);
-  size_t total_values = std::max(num_def_levels, *values_read);
-  num_decoded_values_ += total_values;
 
+    current_values_read = ReadValues(values_to_read, values + value_offset);
+    size_t current_values = std::max(num_def_levels, current_values_read);
```

Comment:
This works in the case when definition levels are absent too. Did you mean to have two assignments ? one after reading values and one after reading definition levels?